### PR TITLE
feat: animator play backwards

### PR DESCRIPTION
**Pull request by luzhuang**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

File: packages/core/src/animation/internal/AnimatorStatePlayData.ts [hunk 1]
```diff
@@ -28,12 +28,25 @@ export class AnimatorStatePlayData {
...
+    if (time < 0) {
+      time = duration+ time;
+      if (time < 0) {
+        if (state.wrapMode === WrapMode.Loop) {
+          time = duration + (time % duration);
+        } else {
+          time = 0;
+          this.playState = AnimatorStatePlayState.Finished;
+        }
       }
```
Comment (GuoLei1990) [inline packages/core/src/animation/internal/AnimatorStatePlayData.ts:50]: Simplified logic like this:
**Important: When adding features, you should re-integrate and simplify the logic**
```ts
if (state.wrapMode === WrapMode.Loop) {
      time = time % duration;
      time < 0 && (time += duration);
    } else {
      if (Math.abs(time) > duration) {
        time = duration;
        this.playState = AnimatorStatePlayState.Finished;
      }
    }
```

File: packages/core/src/animation/Animator.ts [hunk 2]
```diff
@@ -872,6 +881,30 @@ export class Animator extends Component {
...
+      const { handlers } = eventHandler;
+      if (time <= lastClipTime) {
+        for (let j = handlers.length - 1; j >= 0; j--) {
+          handlers[j](parameter);
+        }
+        playState.currentEventIndex = i + 1;
+      }
+    }
+  }
+
```
Comment (GuoLei1990) [inline packages/core/src/animation/Animator.ts:907]: - reduce assignment of `playState.currentEventIndex`;
- Use more nature loop logic to optimization `_fireBackwardSubAnimationEvents`

```ts
private _fireSubAnimationEvents(
    playState: AnimatorStatePlayData,
    eventHandlers: AnimationEventHandler[],
    lastClipTime: number,
    curClipTime: number
  ): void {
    let eventIndex = playState.currentEventIndex;
    for (let n = eventHandlers.length; eventIndex < n; eventIndex++) {
      const eventHandler = eventHandlers[eventIndex];
      const { time, parameter } = eventHandler.event;

      if (time > curClipTime) {
        break;
      }

      const { handlers } = eventHandler;
      if (time >= lastClipTime) {
        for (let j = handlers.length - 1; j >= 0; j--) {
          handlers[j](parameter);
        }
      }
    }
    playState.currentEventIndex = eventIndex;
  }

  private _fireBackwardSubAnimationEvents(
    playState: AnimatorStatePlayData,
    eventHandlers: AnimationEventHandler[],
    lastClipTime: number,
    curClipTime: number
  ): void {
    let eventIndex = playState.currentEventIndex;
    for (; eventIndex >= 0; eventIndex--) {
      const eventHandler = eventHandlers[eventIndex];
      const { time, parameter } = eventHandler.event;

      if (time < curClipTime) {
        break;
      }

      const { handlers } = eventHandler;
      if (time <= lastClipTime) {
        for (let j = handlers.length - 1; j >= 0; j--) {
          handlers[j](parameter);
        }
      }
    }
    playState.currentEventIndex = eventIndex;
  }
```

**Review by zhuxudong — APPROVED** [review]
+1

File: packages/core/src/animation/internal/AnimatorStatePlayData.ts (hunk 1 above)
Comment (luzhuang) [reply packages/core/src/animation/internal/AnimatorStatePlayData.ts:39]: > Simplified logic like this: **Important: When adding features, you should re-integrate and simplify the logic**
> 
> ```ts
> if (Math.abs(time) > duration) {
>  time = duration;
>}
> ```
When wrapMode !== WrapMode.Loop, i think time should be 0

File: packages/core/src/animation/Animator.ts (hunk 2 above)
Comment (luzhuang) [reply packages/core/src/animation/Animator.ts]: > ```ts
>     playState.currentEventIndex = eventIndex;
> ```
When trigger break, i will be wrong.

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**